File: college/fppd/ex_04/ex_04.py
# Cumulative sum of previous elements
from threading import Thread
import time
import random

def prefix_sum_serial(A):
    B = [0] * len(A)
    B[0] = A[0]
    for i in range(1, len(A)):
        B[i] = B[i - 1] + A[i]
    return B

def prefix_sum_thread(A, B, start, end):
    B[start] = A[start]
    for i in range(start + 1, end):
        B[i] = B[i - 1] + A[i]

def prefix_sum_with_threads(A, num_threads=4):
    n = len(A)
    B = [0] * n
    threads = []
    chunk_size = n // num_threads

    for i in range(num_threads):
        start = i * chunk_size
        end = n if i == num_threads - 1 else (i + 1) * chunk_size
        thread = Thread(target=prefix_sum_thread, args=(A, B, start, end))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    # Adjust global prefixes
    for i in range(1, num_threads):
        start = i * chunk_size
        offset = B[start - 1]
        for j in range(start, min((i + 1) * chunk_size, n)):
            B[j] += offset

    return B

def main():
    # Generate the array A with 50,000 random numbers
    A = [random.randint(1, 500000) for _ in range(50000)]

    # Serial computation
    start_time_serial = time.time()
    B_serial = prefix_sum_serial(A)
    time_serial = time.time() - start_time_serial

    # Parallel computation
    start_time_parallel = time.time()
    B_parallel = prefix_sum_with_threads(A)
    time_parallel = time.time() - start_time_parallel

    # Check if results are the same
    assert B_serial == B_parallel, "The serial and parallel results are different!"

    # Calculate speedup
    speedup = time_serial / time_parallel

    # Display the results
    print(f"Serial time: {time_serial:.5f} seconds")
    print(f"Parallel time: {time_parallel:.5f} seconds")
    print(f"Speedup: {speedup:.2f}")
    print("\nFirst 10 elements of vector A:", A[:10])
    print("\nFirst 10 elements of vector BSerial:", B_serial[:10])
    print("\nFirst 10 elements of vector BParallel:", B_parallel[:10])
    assert B_parallel == B_serial, "Error: Results do not match!"

# Run the program
if __name__ == "__main__":
    main()

File: college/fppd/ex_04/ex_04.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <omp.h>


void prefixSumSerial(int* A, int* B, int n) {
    B[0] = A[0];
    for (int i = 1; i < n; i++) {
        B[i] = B[i - 1] + A[i];
    }
}

// Parallel version using OpenMP
void prefixSumParallel(int* A, int* B, int n, int numThreads) {
    int chunkSize = n / numThreads;

    // Step 1: Compute prefix sum for each chunk in parallel
    #pragma omp parallel num_threads(numThreads)
    {
        int tid = omp_get_thread_num();
        int start = tid * chunkSize;
        int end = (tid == numThreads - 1) ? n : (tid + 1) * chunkSize;
        
        // Compute prefix sum for each chunk
        if (start == 0) {
            B[start] = A[start];
        } else {
            B[start] = A[start] + B[start - 1];
        }
        
        for (int i = start + 1; i < end; i++) {
            B[i] = B[i - 1] + A[i];
        }
    }

    // Step 2: Adjust each chunk based on the previous chunk's last value
    #pragma omp parallel for
    for (int i = 1; i < numThreads; i++) {
        int adjustValue = B[i * chunkSize - 1];
        for (int j = i * chunkSize; j < ((i + 1) * chunkSize < n ? (i + 1) * chunkSize : n); j++) {
            B[j] += adjustValue;
        }
    }
}

int main() {
    // Array size and number of threads
    int n = 50000;
    int numThreads = 4;

    // Initialize the array A with random values
    int* A = (int*)malloc(n * sizeof(int));
    int* BSerial = (int*)malloc(n * sizeof(int));
    int* BParallel = (int*)malloc(n * sizeof(int));

    srand(time(NULL));
    for (int i = 0; i < n; i++) {
        A[i] = rand() % 10000;
    }

    // Serial execution
    clock_t startTime = clock();
    prefixSumSerial(A, BSerial, n);
    clock_t endTime = clock();
    double serialTime = (double)(endTime - startTime) / CLOCKS_PER_SEC;
    printf("Serial execution time: %.5f seconds\n", serialTime);

    // Parallel execution
    startTime = clock();
    prefixSumParallel(A, BParallel, n, numThreads);
    endTime = clock();
    double parallelTime = (double)(endTime - startTime) / CLOCKS_PER_SEC;
    printf("Parallel execution time: %.5f seconds\n", parallelTime);

    // Calculate speedup
    double speedup = serialTime / parallelTime;
    printf("Speedup: %.2f\n", speedup);

    // Free allocated memory
    free(A);
    free(BSerial);
    free(BParallel);

    return 0;
}
